File: src/types/events.ts
import type { CharacterData } from './character'
import type { ChatMessage } from './api'

// 角色加载事件载荷
export interface CharacterLoadedPayload {
  uuid: string
  character_data: CharacterData
  timestamp: number
}

// 聊天历史加载事件载荷
export interface ChatHistoryLoadedPayload {
  uuid: string
  chat_history: ChatMessage[]
  message_count: number
  timestamp: number
}

// 消息发送事件载荷
export interface MessageSentPayload {
  uuid: string
  message: ChatMessage
  timestamp: number
}

// 消息接收事件载荷
export interface MessageReceivedPayload {
  uuid: string
  message: ChatMessage
  timestamp: number
  /** 中间消息（包括 assistant with tool_calls 和 tool results） */
  intermediate_messages?: ChatMessage[]
}

// 上下文构建完成事件载荷
export interface ContextBuiltPayload {
  uuid: string
  context_result: BuiltContextResult
  timestamp: number
}

// 角色更新事件载荷
export interface CharacterUpdatedPayload {
  uuid: string
  character_data: CharacterData
  update_type: CharacterUpdateType
  timestamp: number
}

// 角色更新类型
export type CharacterUpdateType =
  | 'BasicInfo'
  | 'Worldbook'
  | 'Tags'
  | 'FullData'
  | { Fields: { fields: string[] } }

// 工具执行事件载荷
export interface ToolExecutedPayload {
  uuid: string
  tool_name: string
  success: boolean
  result?: any
  error?: string
  execution_time_ms: number
  timestamp: number
}

// 会话卸载事件载荷
export interface SessionUnloadedPayload {
  uuid: string
  session_info: SessionInfo
  reason: SessionUnloadReason
  timestamp: number
}

// 会话卸载原因
export type SessionUnloadReason =
  | 'UserRequest'
  | 'Expired'
  | 'MemoryCleanup'
  | 'AppShutdown'
  | { Error: string }

// 错误事件载荷
export interface ErrorPayload {
  uuid?: string
  error_code: string
  error_message: string
  error_context?: any
  timestamp: number
}

// Token统计事件载荷
export interface TokenStatsPayload {
  uuid: string
  token_usage: TokenUsageStats
  timestamp: number
}

// Token使用统计
export interface TokenUsageStats {
  prompt_tokens: number
  completion_tokens: number
  total_tokens: number
  context_tokens: number
  budget_utilization: number
}

// 进度事件载荷
export interface ProgressPayload {
  uuid: string
  operation: string
  progress: number // 0.0 - 1.0
  message?: string
  timestamp: number
}

// 会话信息
export interface SessionInfo {
  uuid: string
  character_name?: string
  message_count: number
  last_active: string
  status: SessionStatus
  last_context_tokens: number
}

// 会话状态
export type SessionStatus = 'Active' | 'Paused' | 'Loading' | { Error: string }

// Token分配详情
export interface TokenAllocation {
  character: number
  worldbook: number
  system: number
  history: number
}

// 构建完成的上下文结果
export interface BuiltContextResult {
  system_messages: any[] // OpenAIMessage[]
  assistant_messages: any[] // OpenAIMessage[]
  history_messages: any[] // OpenAIMessage[]
  current_user_message?: any // OpenAIMessage
  total_tokens: number
  token_allocation: TokenAllocation
  was_truncated: boolean
}

// OpenAI消息结构
export interface OpenAIMessage {
  role: string
  content: string
  name?: string
  tool_calls?: any[] // ToolCall[]
  tool_call_id?: string
}

File: src-tauri/src/events.rs
use crate::ai_chat::{MessageRole, ToolCallData};
use crate::backend::domain::{
    CharacterLoadedPayload, CharacterUpdateType, CharacterUpdatedPayload, ChatHistoryLoadedPayload,
    ContextBuiltPayload, MessageReasoningDeltaPayload, MessageReceivedPayload, MessageSentPayload,
    MessageStreamDeltaPayload, ReasoningDeltaKind, SessionInfo, SessionUnloadReason,
    SessionUnloadedPayload, TokenStatsPayload, TokenUsageStats, ToolExecutedPayload,
    ToolExecutionPhase, ToolExecutionStatusPayload,
};
use crate::character_storage::CharacterData;
use crate::chat_history::ChatMessage;
use crate::context_builder::BuiltContextResult;
use tauri::{AppHandle, Emitter};

/// 事件发送器 - 提供统一的事件发送接口
pub struct EventEmitter;

impl EventEmitter {
    /// 发送角色加载事件
    pub fn send_character_loaded(
        app: &AppHandle,
        uuid: &str,
        character_data: &CharacterData,
    ) -> Result<(), String> {
        let payload = CharacterLoadedPayload {
            uuid: uuid.to_string(),
            character_data: character_data.clone(),
            timestamp: chrono::Utc::now().timestamp(),
        };

        app.emit("character-loaded", &payload)
            .map_err(|e| format!("发送角色加载事件失败: {}", e))?;

        Ok(())
    }

    /// 发送聊天历史加载事件
    pub fn send_chat_history_loaded(
        app: &AppHandle,
        uuid: &str,
        chat_history: &[ChatMessage],
    ) -> Result<(), String> {
        let payload = ChatHistoryLoadedPayload {
            uuid: uuid.to_string(),
            chat_history: chat_history.to_vec(),
            message_count: chat_history.len(),
            timestamp: chrono::Utc::now().timestamp(),
        };

        app.emit("chat-history-loaded", &payload)
            .map_err(|e| format!("发送聊天历史加载事件失败: {}", e))?;

        Ok(())
    }

    /// 发送消息发送事件
    pub fn send_message_sent(
        app: &AppHandle,
        uuid: &str,
        message: &ChatMessage,
    ) -> Result<(), String> {
        let payload = MessageSentPayload {
            uuid: uuid.to_string(),
            message: message.clone(),
            timestamp: chrono::Utc::now().timestamp(),
        };

        app.emit("message-sent", &payload)
            .map_err(|e| format!("发送消息事件失败: {}", e))?;

        Ok(())
    }

    /// 发送消息接收事件
    pub fn send_message_received(
        app: &AppHandle,
        uuid: &str,
        message: &ChatMessage,
        target_message_id: Option<&str>,
        intermediate_messages: Option<Vec<ChatMessage>>,
    ) -> Result<(), String> {
        let payload = MessageReceivedPayload {
            uuid: uuid.to_string(),
            message: message.clone(),
            target_message_id: target_message_id.map(|value| value.to_string()),
            timestamp: chrono::Utc::now().timestamp(),
            intermediate_messages,
        };

        app.emit("message-received", &payload)
            .map_err(|e| format!("发送接收消息事件失败: {}", e))?;

        Ok(())
    }

    pub fn send_message_stream_delta(
        app: &AppHandle,
        uuid: &str,
        target_message_id: &str,
        role: MessageRole,
        delta: &str,
        is_finished: bool,
        is_aborted: bool,
    ) -> Result<(), String> {
        let payload = MessageStreamDeltaPayload {
            uuid: uuid.to_string(),
            role,
            target_message_id: target_message_id.to_string(),
            delta: delta.to_string(),
            is_finished,
            is_aborted,
            timestamp: chrono::Utc::now().timestamp(),
        };

        app.emit("message-stream-delta", &payload)
            .map_err(|e| format!("发送流式消息事件失败: {}", e))?;

        Ok(())
    }

    pub fn send_message_reasoning_delta(
        app: &AppHandle,
        uuid: &str,
        target_message_id: &str,
        delta: &str,
        kind: ReasoningDeltaKind,
        is_finished: bool,
        is_aborted: bool,
    ) -> Result<(), String> {
        let payload = MessageReasoningDeltaPayload {
            uuid: uuid.to_string(),
            target_message_id: target_message_id.to_string(),
            delta: delta.to_string(),
            kind,
            is_finished,
            is_aborted,
            timestamp: chrono::Utc::now().timestamp(),
        };

        app.emit("message-reasoning-delta", &payload)
            .map_err(|e| format!("发送 reasoning 流式消息事件失败: {}", e))?;

        Ok(())
    }

    pub fn send_tool_execution_status(
        app: &AppHandle,
        uuid: &str,
        target_message_id: &str,
        tool_call: &ToolCallData,
        phase: ToolExecutionPhase,
        result: Option<serde_json::Value>,
        error: Option<String>,
        execution_time_ms: Option<u64>,
    ) -> Result<(), String> {
        let payload = ToolExecutionStatusPayload {
            uuid: uuid.to_string(),
            target_message_id: target_message_id.to_string(),
            tool_call_id: tool_call.id.clone(),
            tool_name: tool_call.function.name.clone(),
            phase,
            tool_call: Some(tool_call.clone()),
            result,
            error,
            execution_time_ms,
            timestamp: chrono::Utc::now().timestamp(),
        };

        app.emit("tool-execution-status", &payload)
            .map_err(|e| format!("发送工具状态事件失败: {}", e))?;

        Ok(())
    }

    /// 发送上下文构建完成事件
    pub fn send_context_built(
        app: &AppHandle,
        uuid: &str,
        context_result: &BuiltContextResult,
    ) -> Result<(), String> {
        let payload = ContextBuiltPayload {
            uuid: uuid.to_string(),
            context_result: context_result.clone(),
            timestamp: chrono::Utc::now().timestamp(),
        };

        app.emit("context-built", &payload)
            .map_err(|e| format!("发送上下文构建事件失败: {}", e))?;

        Ok(())
    }

    /// 发送角色更新事件
    pub fn send_character_updated(
        app: &AppHandle,
        uuid: &str,
        character_data: &CharacterData,
        update_type: CharacterUpdateType,
    ) -> Result<(), String> {
        let payload = CharacterUpdatedPayload {
            uuid: uuid.to_string(),
            character_data: character_data.clone(),
            update_type,
            timestamp: chrono::Utc::now().timestamp(),
        };

        app.emit("character-updated", &payload)
            .map_err(|e| format!("发送角色更新事件失败: {}", e))?;

        Ok(())
    }

    /// 发送工具执行事件
    pub fn send_tool_executed(
        app: &AppHandle,
        uuid: &str,
        tool_name: &str,
        success: bool,
        result: Option<serde_json::Value>,
        error: Option<String>,
        execution_time_ms: u64,
    ) -> Result<(), String> {
        let payload = ToolExecutedPayload {
            uuid: uuid.to_string(),
            tool_name: tool_name.to_string(),
            success,
            result,
            error,
            execution_time_ms,
            timestamp: chrono::Utc::now().timestamp(),
        };

        app.emit("tool-executed", &payload)
            .map_err(|e| format!("发送工具执行事件失败: {}", e))?;

        Ok(())
    }

    /// 发送会话卸载事件
    pub fn send_session_unloaded(
        app: &AppHandle,
        uuid: &str,
        session_info: &SessionInfo,
        reason: SessionUnloadReason,
    ) -> Result<(), String> {
        let payload = SessionUnloadedPayload {
            uuid: uuid.to_string(),
            session_info: session_info.clone(),
            reason,
            timestamp: chrono::Utc::now().timestamp(),
        };

        app.emit("session-unloaded", &payload)
            .map_err(|e| format!("发送会话卸载事件失败: {}", e))?;

        Ok(())
    }

    /// 发送Token统计事件
    pub fn send_token_stats(
        app: &AppHandle,
        uuid: &str,
        token_usage: TokenUsageStats,
    ) -> Result<(), String> {
        let payload = TokenStatsPayload {
            uuid: uuid.to_string(),
            token_usage,
            timestamp: chrono::Utc::now().timestamp(),
        };

        app.emit("token-stats", &payload)
            .map_err(|e| format!("发送Token统计事件失败: {}", e))?;

        Ok(())
    }

    /// 发送通用进度事件（用于长时间操作）
    pub fn send_progress(
        app: &AppHandle,
        uuid: &str,
        operation: &str,
        progress: f64, // 0.0 - 1.0
        message: Option<&str>,
    ) -> Result<(), String> {
        let payload = serde_json::json!({
            "uuid": uuid,
            "operation": operation,
            "progress": progress,
            "message": message,
            "timestamp": chrono::Utc::now().timestamp()
        });

        app.emit("progress", &payload)
            .map_err(|e| format!("发送进度事件失败: {}", e))?;

        Ok(())
    }
}
